### PR TITLE
sendMediaGroup API operation

### PR DESCRIPTION
**Pull request by vjgarciag96**
This PR fixes the problems reported on #45

The `sendMediaGroup` operation will allow to send a group of media (from 2 to 10 photos and/or videos) as an album to any chat

The media elements composing the group can be sent as file ids, file urls or files (with the possibility of combining all these types of files)

There's an `InputMedia` model defined by Telegram that represents any type of media that can be sent to a chat. This `InputMedia` is not limited to only photos or videos but
also other media types like audio, documents, etc... So, to represent the `InputMedia` model we have a sealed class, and for every concretion (photo, video, etc..) we have a class extending it. In this case, as the type of media that can be sent as media groups is limited to photos and videos, a new interface called `GroupableMedia` has been created to mark all the media types that can be sent as a group and the `InputMediaPhoto` and `InputMediaVideo` classes have been marked with it.

A few tests have been added for the `MediaGroup` class, to check that its validation is working as expected and for the new API operation, where we have added some integration tests where we check that the multipart request body that we send to Telegram's servers is correct.

Also, an example of the usage of this operation has been added to the dispatcher sample

### CODE REVIEW

File: telegram/src/main/kotlin/com/github/kotlintelegrambot/network/serialization/GsonFactory.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+    fun createForApiClient(): Gson = GsonBuilder()
+        .registerTypeAdapter(InlineQueryResult::class.java, InlineQueryResultAdapter())
+        .create()
+
+    fun createForMultipartBodyFactory(): Gson = GsonBuilder()
+        .registerTypeAdapter(TelegramFile.ByFile::class.java, TelegramFileAdapter())
+        .registerTypeAdapter(TelegramFile::class.java, TelegramFileAdapter())
+        .registerTypeAdapter(GroupableMedia::class.java, GroupableMediaAdapter(InputMediaAdapter()))
+        .create()
```
Comment (vjgarciag96) [inline telegram/src/main/kotlin/com/github/kotlintelegrambot/network/serialization/GsonFactory.kt:23]: I've created different factory methods here to create Gson instances with different configurations to only have the adapters needed in each component and not having all of them in all the instances

File: telegram/src/main/kotlin/com/github/kotlintelegrambot/network/serialization/adapter/TelegramFileAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import com.google.gson.JsonSerializationContext
+import com.google.gson.JsonSerializer
+import java.lang.reflect.Type
+
+class TelegramFileAdapter : JsonSerializer<TelegramFile> {
+    override fun serialize(src: TelegramFile, typeOfSrc: Type, context: JsonSerializationContext): JsonElement = when (src) {
+        is ByFileId -> context.serialize(src.fileId, String::class.java)
+        is ByUrl -> context.serialize(src.url, String::class.java)
+        is ByFile -> context.serialize("attach://${src.file.name}")
+    }
```
Comment (vjgarciag96) [inline telegram/src/main/kotlin/com/github/kotlintelegrambot/network/serialization/adapter/TelegramFileAdapter.kt:17]: This is the way the different types of file must be serialized for the `sendMediaGroup` operation

`attach://filename` indicates that a file with that name is being sent as part of the multipart request body (it's the approach Telegram has stablished for this case)

File: telegram/src/main/kotlin/com/github/kotlintelegrambot/network/multipart/MultipartExtensions.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package com.github.kotlintelegrambot.network.multipart
```
Comment (vjgarciag96) [inline telegram/src/main/kotlin/com/github/kotlintelegrambot/network/multipart/MultipartExtensions.kt:1]: Just moved

File: telegram/src/main/kotlin/com/github/kotlintelegrambot/entities/inputmedia/InputMedia.kt [hunk 4]
```diff
@@ -1,7 +1,112 @@
...
+    abstract val parseMode: String?
+}
+
+/**
+ * Interface to mark all the media types that can be sent within a group of media for
+ * operations like `sendMediaGroup`.
+ */
+interface GroupableMedia
+
+class MediaGroup private constructor(val medias: Array<out GroupableMedia>) {
```
Comment (manolovn) [inline telegram/src/main/kotlin/com/github/kotlintelegrambot/entities/inputmedia/InputMedia.kt:23]: a Set instead of Array could helps here?

Comment (vjgarciag96) [reply telegram/src/main/kotlin/com/github/kotlintelegrambot/entities/inputmedia/InputMedia.kt:23]: hmm... the point is that you can send the same media two times, I guess that's not the usual use case, but using a set wouldn't be accurate enough for this reason. 

Also, using an `Array` lets us taking advantage of `varargs` in an easier way

**Pull request merged**